File: State/java/State.java
abstract class State{
    protected TrafficLight trafficLight;

    public void setTrafficLight(TrafficLight trafficLight){
        this.trafficLight = trafficLight;
    }

    public abstract void nextState();

    public abstract void previousState();
}

class TrafficLight{
    private State state;

    public TrafficLight(State st){
        setState(st);
    }

    public void setState(State st){
        state = st;
        state.setTrafficLight(this);
    }

    public void nextState(){
        state.nextState();
    }

    public void previousState(){
        state.previousState();
    }
}

class GreenState extends State{
    @Override
    public void nextState(){
        System.out.println("from green to yellow");
        trafficLight.setState(new YellowState());
    }

    @Override
    public void previousState(){
        System.out.println("green color");
    }
}

class YellowState extends State{
    @Override
    public void nextState(){
        System.out.println("from yellow to red");
        trafficLight.setState(new RedState());
    }

    @Override
    public void previousState(){
        System.out.println("from yellow to green");
        trafficLight.setState(new GreenState());
    }
}

class RedState extends State{
    @Override
    public void nextState(){
        System.out.println("red color");
    }
    
    @Override
    public void previousState(){
        System.out.println("from red to yellow");
        trafficLight.setState(new YellowState());
    }
}

public class Main {
    public static void main(String[] args){
        TrafficLight trafficLight = new TrafficLight(new GreenState());

        trafficLight.nextState();
        trafficLight.nextState();
        trafficLight.nextState();
        trafficLight.previousState();
        trafficLight.previousState();
        trafficLight.previousState();
    }
}

File: State/c#/State.cs
abstract class State
{
    protected TrafficLight _trafficLight;
    public TrafficLight TrafficLight { set => _trafficLight = value; }
    public abstract void NextState();
    public abstract void PreviousState();
}

class TrafficLight
{
    private State _state;
    
    public TrafficLight(State st) => SetState(st);

    public void SetState(State st)
    {
        _state = st;
        _state.TrafficLight = this;
    }

    public void NextState()
    {
        _state.NextState();
    }

    public void PreviousState()
    {
        _state.PreviousState();
    }
}

class GreenState : State
{
    public override void NextState()
    {
        Console.WriteLine("from green to yellow");
        _trafficLight.SetState(new YellowState());
    }

    public override void PreviousState()
    {
        Console.WriteLine("Green color");
    }
}

class YellowState : State
{
    public override void NextState()
    {
        Console.WriteLine("from yellow to red");
        _trafficLight.SetState(new RedState());
    }

    public override void PreviousState()
    {
        Console.WriteLine("from yellow to green");
        _trafficLight.SetState(new GreenState());
    }
}

class RedState : State
{
    public override void NextState()
    {
        Console.WriteLine("red color");
    }

    public override void PreviousState()
    {
        Console.WriteLine("from red to yellow");
        _trafficLight.SetState(new YellowState());
    }
}

class Program
{
    static void Main()
    {
        TrafficLight trafficLight = new TrafficLight(new GreenState());
        
        trafficLight.NextState();
        trafficLight.NextState();
        trafficLight.NextState();
        trafficLight.PreviousState();
        trafficLight.PreviousState();
        trafficLight.PreviousState();
    }
}